Create new `base-cms-theme-monorail` package to replace tenant based versions

B77Mills opened the pull request:
Create a latest version of this package that is a roll up of the ones currently being used in the individual tenant repos.  This new shared package with replace the current on in the given tenant't packages.

B77Mills committed:
Convert monorail package from latest RR version

B77Mills committed:
add latest changes to monorail theme

B77Mills committed:
fix browser lint file

solocommand commented on packages/marko-web-theme-monorail/components/marko.json:
```diff
@@ -0,0 +1,40 @@
+    "./nodes/marko.json",
+    "./page/marko.json"
+  ],
+  "<theme-client-side-block-loader>": {
+    "template": "./client-side-block-loader.marko"
+  },
+  "<theme-menu-toggle-button>": {
+    "template": "./menu-toggle-button.marko"
+  },
+  "<theme-pagination-controls>": {
+    "template": "./pagination-controls.marko"
+  },
+  "<theme-recaptcha>": {
+    "template": "./recaptcha.marko"
+  },
+  "<theme-site-footer>": {
+    "template": "./site-footer.marko"
+  },
+  "<theme-site-menu>": {
+    "template": "./site-menu/index.marko"
+  },
+  "<theme-site-newsletter-menu>": {
+    "template": "./site-newsletter-menu.marko"
+  },
+  "<theme-social-follow>": {
+    "template": "./social-follow.marko"
+  },
+  "<theme-query-total-count>": {
+    "template": "./query-total-count.marko"
+  }
```
Make sure every component has a defined list of inputs. The only exception to this rule is for structural issues arising from the need to spread objects as input (such as the global document component). There should be only a handful of these across the entire project.

This applies to the other `marko.json` files in subfolders as well, but I'm not going to comment on them individually.

This is primarily to ensure that the components are being accessed consistently, and to verify that changes made to the package do not cause backwards-incompatible breaks in dependent projects.

solocommand commented on packages/marko-web-theme-monorail/components/menu-toggle-button.marko:
```diff
@@ -0,0 +1,19 @@
+    beforeCollapsed: input.beforeCollapsed,
+    after: input.after,
+    afterExpanded: input.afterExpanded,
+    afterColapsed: input.afterColapsed,
```
typo, use `afterCollapsed`. Update Vue component property as well

solocommand commented on packages/marko-web-theme-monorail/browser/block-loader.vue:
```diff
@@ -0,0 +1,64 @@
+    <div v-if="isLoading">
+      Loading {{ label }}...
+    </div>
+    <!-- eslint-disable-next-line vue/no-v-html -->
```
Add this exception to your eslintrc.js file instead of an inline disable. See the leaders-program package for an example of this.

solocommand commented on packages/marko-web-theme-monorail/browser/comment-toggle-button.vue:
```diff
@@ -0,0 +1,87 @@
+
+    toggle() {
+      this.expanded = !this.expanded;
+      EventBus.$emit('comments-expanded', this.expanded);
```
This _should_ be prefixed with `identity-x-`. Before changing, verify that no user-land events are bound to it (GTM or site/browser stuff). Also/or add an `@todo` to move this functionality to the `marko-web-identity-x` package in the future.

solocommand commented on packages/marko-web-theme-monorail/browser/top-stories-menu.vue:
```diff
@@ -0,0 +1,58 @@
+    <div v-if="isLoading">
+      Loading top stories...
+    </div>
+    <!-- eslint-disable-next-line vue/no-v-html -->
```
This should be covered by the eslintrc change and can be removed

solocommand commented on packages/marko-web-theme-monorail/components/blocks/most-popular.marko:
```diff
@@ -0,0 +1,85 @@
+$ const blockName = "most-popular";
+
+$ const getMostPopular = async () => {
+  const uri = MOST_POPULAR_API_URI || 'https://most-popular.caprica.base.parameter1.com';
```
This should be `most-popular-content.base.parameter1.com`

solocommand commented on packages/marko-web-theme-monorail/components/blocks/most-popular.marko:
```diff
@@ -0,0 +1,85 @@
+  const url = `${uri}/retrieve?tenant=${tenant}&realm=${realm}`;
+  const res = await fetch(url);
+  const json = await res.json();
+  if (!res.ok) throw new Error(json.message || res.statusText || "An unknown fatal error occurred.");  return json.data;
```
don't inline this, just use the braced format

solocommand commented on packages/marko-web-theme-monorail/components/blocks/most-popular.marko:
```diff
@@ -0,0 +1,85 @@
+          </default-theme-card-deck-flow>
+        </marko-web-block>
+      </if>
+    </marko-web-resolve>
+  </if>
+</marko-web-resolve>
```
Something is off with your closing tags and if structure here

B77Mills replied:
@solocommand I dont think there is anything off with the tags.  The closing part looks odd cause its structured like 
```
<marko-web-resolver>(getMostPopularIds)>
  <if(ids>
    <marko-web-resolver>(getContentFromIds)>
        $ const nodes = getAsArray(response, "data.allContent.edges").map((edge) => edge.node);
        $ const map = nodes.reduce((m, r) => m.set(`${r.id}`, r), new Map());
        $ const ordered = ids.map(id => map.get(`${id}`)).filter(id => id);
        <if(ordered.length)>
        <marko-web-block name=blockName>
      <if(ordered>
      </if>
    </marko-web-resolver>
  </if>
</marko-web-resolver>
```

solocommand commented on packages/marko-web-theme-monorail/components/blocks/newsletter-signup-banner-large.marko:
```diff
@@ -0,0 +1,30 @@
+$ const blockName = "newsletter-signup-banner-large";
+
+$ const { site } = out.global;
+$ const { action, hiddenInputs, name, description, disabled } = site.getAsObject('newsletter.signupBannerLarge');
```
Break to new lines for more than 3 props

solocommand commented on packages/marko-web-theme-monorail/browser/inline-newsletter-form.vue:
```diff
@@ -0,0 +1,129 @@
+      this.$emit('submit', { step: 1 });
+      this.email = email;
+      this.step = 2;
+      if (window.olytics) window.olytics.confirm(encryptedCustomerId);
```
Don't run this inside the component; add this logic to your `submit` event binding in the browser.js file (around L97). This is done in another component as well, move it from there too

B77Mills replied:
@solocommand I moved line 125 into the index.js login with the following changes. and made the same change to site-newsletter-menu.vue as well.  

```
Browser.register('ThemeInlineNewsletterForm', InlineNewsletterForm, {
    on: {
      load: data => emitNewsletterEvent({ type: 'Inline', action: 'Load', data }),
      view: data => emitNewsletterEvent({ type: 'Inline', action: 'View', data }),
      focus: data => emitNewsletterEvent({ type: 'Inline', action: 'Focus', data }),
      ###
      submit: (data) => {
        emitNewsletterEvent({ type: 'Inline', action: 'Submit', data });
        if (window.olytics) window.olytics.confirm(data.encryptedCustomerId);
      },
      ###
      subscribe: ({ newsletter }) => emitNewsletterSubscription({ type: 'Pushdown', newsletter }),
      error: data => emitNewsletterEvent({ type: 'Inline', action: 'Error', data: { ...data, error: data.error.message } }),
    },
  });
  ```

solocommand reviewed (changes requested):
@B77Mills more to come, but the biggest issue is a broken Marko component and lack of input definitions in your marko.json files

B77Mills committed:
fix spelling error

B77Mills committed:
Set proper uri for most popular api

B77Mills committed:
Add list of missing props to marko.json files

B77Mills committed:
prepend identity-x- to event name

B77Mills committed:
cleanup destructure of object

B77Mills committed:
move shared window.olytics.confirm on submit to index.js

Move the window.olytics.confirm(data.enctryptedCustomerId) to index.js

B77Mills committed:
cleanup most popular error message handeling

B77Mills committed:
Fix/Cleanup Step 2 of newsletter sign up.

B77Mills committed:
Pass privacy policy to step 2 for inline newsletter forms

Missed  adding privacy policy to inline form when I added it to the standard form

B77Mills committed:
Ad remaining missing props

B77Mills committed:
Dont know why i spread it

B77Mills committed:
Revert "Ad remaining missing props"

This reverts commit 4f0c81eded511faa93b352fab4971aea87d4e291.

B77Mills committed:
Revert "Dont know why i spread it"

This reverts commit 25a984bab1255389f6536d69e6908470c4d77040.

B77Mills committed:
Revert "Add list of missing props to marko.json files"

This reverts commit 708fe69d27a2cf7e6c7e8043aa46eaebc0c8dc1f.

B77Mills committed:
revert this back to what it was before

solocommand commented on packages/marko-web-theme-monorail/components/gam/components/define-display-ad.marko:
```diff
@@ -0,0 +1,102 @@
+import { BEM, warn } from "@parameter1/base-cms-utils";
```
add an `@todo` here: Merge this component into marko-web-gam's define-display-ad. Bring over the utility function and add the referrer and position targeting.

This should have been done previously instead of making a custom component for it, but will let go for now.

solocommand commented on packages/marko-web-theme-monorail/components/gam/components/prestitial-script.marko:
```diff
@@ -0,0 +1,29 @@
+<script>
```
Remove all references to prestitial from the theme

solocommand commented on packages/marko-web-theme-monorail/components/gam/define-display-ad.marko:
```diff
@@ -0,0 +1,25 @@
+];
+$ if (templateName) modifiers.push(`template-${dasherize(templateName)}`);
+
+<define-display-ad
```
Add `@todo`: replace with `<marko-web-gam-define-display-ad>` once logic is referrer and position targeting is migrated

solocommand commented on packages/marko-web-theme-monorail/components/gam/inject-ads.marko:
```diff
@@ -0,0 +1,23 @@
+  delete i.at;
+  i.position = defaultValue(input.position, "content_body");
+  // Ensure closing HMTL tags are escaped.
+  const html = DefineDisplayAd.renderToString({ ...i, $global: out.global }).replace(/<\/(.+?)>/g, '<\\/$1>');
```
Add `@todo`: Merge this with `<marko-web-gam-inject-ads>`. The `$global: out.global` trick is providing access to the express request and response objects for retrieving the referrer and incrementing/storing the position counter. Investigate if this can be done better -- perhaps passing the referrer directly, along with the res.local prop (or a reference to it).

This line and the default value for `position` are the only differences.

solocommand commented on packages/marko-web-theme-monorail/components/gam/prestitial.marko:
```diff
@@ -0,0 +1,40 @@
+import { getAsObject } from "@parameter1/base-cms-object-path";
```
Remove

solocommand commented on packages/marko-web-theme-monorail/components/gam/utils/increment-pos.js:
```diff
@@ -0,0 +1,11 @@
+module.exports = ({ res, position, inc = 1 } = {}) => {
```
Add `@todo`: Move this logic into the `<marko-web-gam-inject-ads>` component when consolidating.

solocommand commented on packages/marko-web-theme-monorail/components/identity-x/comment-stream.marko:
```diff
@@ -0,0 +1,30 @@
+import { getAsObject } from "@parameter1/base-cms-object-path";
```
Add `@todo`: Move to <marko-web-identity-x-comment-stream>` component. Use toggler based on input (default false)

solocommand commented on packages/marko-web-theme-monorail/components/recaptcha.marko:
```diff
@@ -0,0 +1,5 @@
+
+$ const src = `https://www.google.com/recaptcha/api.js?render=${recaptcha.siteKey}`;
+
+<script src=src></script>
```
Add `@todo`: Test if defer/async will work here. Also review version loaded with vue components for conflict/unification.

solocommand commented on packages/marko-web-theme-monorail/components/site-menu/index.marko:
```diff
@@ -0,0 +1,106 @@
+    <!-- user section || IDX_NAV_ENABLE -->
+    <if(site.getAsArray("navigation.mobileMenu.user"))>
+      <!-- user section -->
+        <default-theme-site-menu-section
```
Remove references to default theme components -- create if they don't already exist in this theme.

solocommand replied:
There seem to be ~30 references to default theme components that need to be updated

solocommand commented on packages/marko-web-theme-monorail/marko.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Can't comment on it, but remove empty file packages/marko-web-theme-monorail/index.js

solocommand commented on packages/marko-web-theme-monorail/package.json:
```diff
@@ -0,0 +1,50 @@
+    "@parameter1/base-cms-utils": "^2.75.1",
+    "@parameter1/base-cms-web-cli": "^2.81.0",
+    "@parameter1/base-cms-web-common": "^2.80.0",
+    "@parameter1/omeda-graphql-client-express": "^0.4.2",
```
Unused. It does look like the step1/step2 vue components are directly accessing the `__omeda/` endpoint -- therefore the initialization of the marko-web-omeda package should likely move into this theme. The direct dependency on the graphql client can be removed here, however.

solocommand commented on packages/marko-web-theme-monorail/components/blocks/white-papers.marko:
```diff
@@ -0,0 +1,74 @@
+import { get } from 'object-path';
```
Swap to the base-cms package and remove the direct `object-path` dependency
```suggestion
import { get } from "@parameter1/base-cms-object-path";
```

solocommand commented on packages/marko-web-theme-monorail/package.json:
```diff
@@ -0,0 +1,50 @@
+    "newrelic": "^6.4.1",
+    "node-fetch": "^2.6.1",
+    "object-path": "^0.11.5",
+    "slug": "^4.0.2"
```
Consolidating comments: The default theme package should be removed, along with direct dependencies for `web-cli` `omeda-graphql-client-express`, and other third-party packages (`body-parser`, `express`, `graphql`, `html-entities`, `newrelic`, `object-path`, and `slug`).

```suggestion
    "@parameter1/base-cms-object-path": "^2.75.1",
    "@parameter1/base-cms-utils": "^2.75.1",
    "@parameter1/base-cms-web-common": "^2.80.0",
    "cheerio": "^1.0.0-rc.10",
    "graphql-tag": "^2.10.1",
    "http-errors": "^1.7.3",
    "moment": "^2.29.1",
    "node-fetch": "^2.6.1"
```

solocommand commented on packages/marko-web-theme-monorail/scss/_functions.scss:
```diff
@@ -0,0 +1,29 @@
+@import "../../node_modules/@parameter1/base-cms-marko-web-theme-default/scss/functions";
```
Move to this theme, don't import.

solocommand commented on packages/marko-web-theme-monorail/scss/components/_spec-guide.scss:
```diff
@@ -0,0 +1,37 @@
+
+// Require Bootstrap tables and input groups
+@import "../../node_modules/bootstrap/scss/tables";
+@import "../../node_modules/bootstrap/scss/input-group";
```
This requires adding bootstrap@`4.3.1` as a dependency in the theme `package.json`

solocommand commented on packages/marko-web-theme-monorail/scss/core.scss:
```diff
@@ -0,0 +1,59 @@
+@import "functions";
+@import "variables";
+
+@import "../../node_modules/@parameter1/base-cms-marko-web-theme-default/scss/theme";
+@import "../../node_modules/bootstrap/scss/spinners";
+
+@import "../../node_modules/@parameter1/base-cms-marko-web-social-sharing/scss/buttons";
+@import "../../node_modules/bootstrap/scss/input-group";
+@import "../../node_modules/@parameter1/base-cms-marko-web-html-sitemap/scss/index";
+
```
Naughty! Remove the default theme references entirely. If you need that theme file, re-create it in this package.

Note the bootstrap comment above, this is required for the input-group and spinners imports. Remove the import for input-group from the other file -- don't include it twice.

If you need to import the css for the sitemap and social sharing packages, make sure they are in your package.json as well.

B77Mills committed:
remove prestitial ad

B77Mills committed:
Add @todo: Merge with marko-web-gam package.

B77Mills committed:
Update packages/marko-web-theme-monorail/components/blocks/white-papers.marko

Co-authored-by: Josh Worden <solocommand@gmail.com>

B77Mills committed:
Removing emtpy index.js file

B77Mills committed:
@todo: defer recaptcha & review  version in use

B77Mills committed:
Add at todo for identityX component merge and ref update

B77Mills committed:
remove duplicate inclusion of css file

B77Mills committed:
remove unused depencencys and add bootstrap 4.3.1

solocommand commented on packages/marko-web-theme-monorail/package.json:
```diff
@@ -0,0 +1,42 @@
+  "repository": "https://github.com/parameter1/base-cms/tree/master/packages/marko-web-theme-monorail",
+  "license": "MIT",
+  "main": "index.js",
+  "private": true,
```
```suggestion
```

solocommand committed:
Update packages/marko-web-theme-monorail/package.json
